cache-prefetch-metrics.go

```diff
@@ -0,0 +1,135 @@
+package rdns
+
+import (
+	"github.com/miekg/dns"
+)
+
+type CachePrefetchMetrics struct {
+	// Cache hit count.
+	items map[CachePrefetchKey]*CachePrefetchEntry
+	// TODO item limit param
+	maxItems int
+	errorCountMax int16
+	hitMin int64
+}
+
+// 0 for no prefetching job
+// 1 for prefetching job is active
+// 2 for stopped by error
+type PrefetchState int
+
+const (
+	PrefetchStateNone   = iota
+	PrefetchStateActive = 1
+	PrefetchStateOther  = 2
+)
+
+type CachePrefetchKey struct {
+	qname string
+	qtype string
+}
+
+type CachePrefetchEntry struct {
+	// request hit count.
+	hit int64
+
+	prefetchState PrefetchState
+	// msg to refetch
+	msg *dns.Msg
+	//ttl      int
+	// fetching error count for discarding error prone fetches
+	errorCount int16
+	key        CachePrefetchKey
+}
+
+func NewCachePrefetchEntry(index CachePrefetchKey) *CachePrefetchEntry {
+	return &CachePrefetchEntry{
+		hit:           0,
+		prefetchState: PrefetchStateNone,
+		msg:           nil,
+		errorCount:    0,
+		key:           index,
+	}
+}
+
+func NewCachePrefetchMetrics(capacity int, errorCountMax int16, hitMin int64) CachePrefetchMetrics {
+		return CachePrefetchMetrics{
+		maxItems: capacity,
+		items:    make(map[CachePrefetchKey]*CachePrefetchEntry),
+		errorCountMax: errorCountMax,
+		hitMin: hitMin,
+
+	}
+}
+func (c *CachePrefetchMetrics) ProcessQuery(query *dns.Msg) {
+	if c.addItem(query) == PrefetchStateNone {
+		c.addHit(query)
+	}
+}
+func (c *CachePrefetchMetrics) addItem(query *dns.Msg) PrefetchState {
+	key := c.getDomainKey(query)
+	item := c.touch(key)
+	if item != nil {
+		return item.prefetchState
+	}
+	// Add new item to the top of the linked list
+	if len(c.items) > c.maxItems {
+		return PrefetchStateOther
+	}
+
+	item = NewCachePrefetchEntry(key)
+	c.items[key] = item
```

Comment:
This looks like it's not safe to be called from multiple goroutines.

Reply:
I see the mutex you added to the caller. Not sure this will be safe with the prefetch loop iterating over the items.

One thing you could try is add the mutex to `CachePrefetchMetrics` and lock/unlock as needed within its methods. That shifts responsibility for locking into CachePrefetchMetrics. But even then you need to be careful with `range r.metrics.items` without a lock.